pkg/agent/proxy/proxier.go

```diff
@@ -130,21 +144,51 @@ func getBindingProtoForIPProto(endpointIP string, protocol corev1.Protocol) bind
 	return bindingProtocol
 }
 
-func (p *proxier) removeStaleEndpoints(staleEndpoints map[k8sproxy.ServicePortName]map[string]k8sproxy.Endpoint) {
-	for svcPortName, endpoints := range staleEndpoints {
-		for _, endpoint := range endpoints {
-			bindingProtocol := getBindingProtoForIPProto(endpoint.IP(), svcPortName.Protocol)
-			if err := p.ofClient.UninstallEndpointFlows(bindingProtocol, endpoint); err != nil {
-				klog.Errorf("Error when removing Endpoint %v for %v", endpoint, svcPortName)
-				continue
-			}
-			if m, ok := p.endpointInstalledMap[svcPortName]; ok {
-				delete(m, endpoint.String())
-				if len(m) == 0 {
-					delete(p.endpointInstalledMap, svcPortName)
+// removeEndpoint removes flows for the given Endpoint from the data path if these flows are no longer
+// needed by any Service. Endpoints from different Services can have the same characteristics and thus
+// can share the same flows. removeEndpoint must be called whenever an Endpoint is no longer used by a
+// given Service. If the Endpoint is still referenced by any other Services, no flow will be removed.
+// The method only returns an error if a data path operation fails. If the flows are successfully
+// removed from the data path, the method returns true. Otherwise, if the flows are still needed for
+// other Services, it returns false.
+func (p *proxier) removeEndpoint(endpoint k8sproxy.Endpoint, protocol binding.Protocol) (bool, error) {
+	key := endpointKey(endpoint, protocol)
+	count := p.endpointReferenceCounter[key]
+	if count == 1 {
+		if err := p.ofClient.UninstallEndpointFlows(protocol, endpoint); err != nil {
+			return false, err
+		}
+		delete(p.endpointReferenceCounter, key)
+		klog.V(2).Infof("Endpoint %s/%s removed", endpoint.String(), protocol)
+	} else if count > 1 {
+		p.endpointReferenceCounter[key] = count - 1
+		klog.V(2).Infof("Stale Endpoint %s/%s is still referenced by other Services, decrementing reference count by 1", endpoint.String(), protocol)
+		return false, nil
+	}
+	return true, nil
+}
+
+// removeStaleEndpoints compares Endpoints we installed with Endpoints we expected. All installed but unexpected Endpoints
+// will be deleted by using removeEndpoint.
+func (p *proxier) removeStaleEndpoints() {
+	hasFailure := false
+	for svcPortName, installedEps := range p.endpointsInstalledMap {
+		for installedEpName, installedEp := range installedEps {
+			if _, ok := p.endpointsMap[svcPortName][installedEpName]; !ok {
+				if _, err := p.removeEndpoint(installedEp, getBindingProtoForIPProto(installedEp.IP(), svcPortName.Protocol)); err != nil {
+					klog.Errorf("Error when removing Endpoint %v for %v", installedEp, svcPortName)
+					hasFailure = true
+					continue
 				}
+				delete(installedEps, installedEpName)
 			}
 		}
+		if len(installedEps) == 0 {
+			delete(p.endpointsInstalledMap, svcPortName)
+		}
+	}
+	if hasFailure {
+		p.runner.Run()
 	}
```

Comment:
@jianjuns @weiqiangt sorry I didn't get why we want to trigger a retry after 1s in case of error, instead of the default 30 seconds? I expect errors to be an edge case, why do we need to retry right away?

Reply:
I did not notice this one. For failures, I agree we should not retry that fast, if we do not have backoff. I was suggesting 10s for now, but keep the default 30s sounds good to me too.

Reply:
Removed.